### ISSUE TITLE
Use pg_hba to disallow direct external access to datanodes, gtm, proxies and coordinators

### ISSUE THREAD

**Issue by LamaAni** [LamaAni/PGXL-HELM#23]
Experimental.

What:

In the current securty context, all database nodes can be accessed via any other host that has access to the kubernetes cluster. In postgres-xl, we can use the pg_hba to only allow connections via the helm chart services, and allow external access only to the main chart service, thus increasing security against in cluster attacks. 

Keywords:
1. Security
2. pg_hba
3. services
4. network

**Comment by stephenstubbs** [comment]
my understanding was that the pg_hba file in postgres xl needs a trust rule for the subnet. This doesn't seem to be the case which is great. I've tested it with md5 and scram-sha-256 and it appears to be working well. 

https://github.com/LamaAni/postgres-xl-helm/pull/35 this is needed to add scram-sha-256 support.

Like you say there is more that can be done for security but this is the reason I added pgbouncer. It doesn't seem as though pgbouncer is needed from a security point of view but can still be useful if people want better performance.

**Comment by LamaAni** [comment]
This ticket is actually on top of pg_bouncer. I consider pg_bouncer to be an added layer of security that can be added to any pg database which is why I think it should be its own chart. 

The reason the subnet is not giving you an error is becuse of the way we apply the security when setting the username, i.e. 
```shell
    initdb \
      -A "${AUTH_TYPE}"\
      -D "${PGDATA}" \
      -U "${PGUSER}" \
      --nodename="${PG_NODE}" \
      --pwfile="${PG_PW_FILE}" || exit $?
```
in  https://github.com/LamaAni/postgres-xl-helm/blob/master/scripts/initialize_postgres_db

And also the network setup, i.e.
```shell
echo "host all all all $PG_AUTH_METHOD" >> "${PGDATA}/pg_hba.conf"
```
in https://github.com/LamaAni/postgres-xl-helm/blob/master/scripts/initialize_networks

As I understand it these remove the subnet dependency.

**Comment by stephenstubbs** [comment]
ok great thank you for the clarification. This is really good news. I will still use pgbouncer as over the years I have seen better performance and memory reduction with my workloads and like you say it does add an extra security layer.

I've been doing some tests with pgbench and sysbench. The easiest tool to setup I've found so far is https://github.com/gregs1104/pgbench-tools. https://emerichunter.github.io/pgbench-tools-p1/ 

Will setup a container with this and let you know what I find.

**Comment by stephenstubbs** [comment]
Should this be closed yet?

**Comment by LamaAni** [comment]
No, I think we have not reached the security level that is appropriate. 

I think we are still allowing all networks instead of just the appropriate pods... I need to think how to do this and keep the deploy stable.

**Comment by stephenstubbs** [comment]
ok makes sense.
